### ISSUE TITLE
Assignment based type inference

### ISSUE THREAD

**Issue by BladeMF** [phpactor/vscode-phpactor#13]
It would be really helpful if type is inferred for variables based on assignment. In that case, because a value with a known type is assigned to `$this->arg1`, and because it has no explicit type declaration, its type should be known.
![image](https://user-images.githubusercontent.com/12109622/102049060-dfb7be00-3de8-11eb-9777-0e3572f5972d.png)

Of course `@var` comments should have the highest precedence.

**Comment by BladeMF** [comment]
I just noticed it workes in method scope:
![image](https://user-images.githubusercontent.com/12109622/102049532-b21f4480-3de9-11eb-9c6e-8e629529dea0.png)

So just needed for fields then.

**Comment by dantleech** [comment]
Yes - this is called constructor property type inference - it was supported for a couple of years but then with the language server it caused an infinite loop and I had to remove it :eyes:

**Comment by BladeMF** [comment]
That is really unfortunate. It's not that I can't live with it, it's just none of my services have type hints :-( Probably with 7.4 that won't be needed?

**Comment by dantleech** [comment]
Yes, I would put it back it's a a question of time and it's not trivial...

**Comment by BladeMF** [comment]
I understand.

**Comment by BladeMF** [comment]
I'd rather have the refactorings implemented than this.
